### PR TITLE
Net: Make net_buf_pool as const

### PR DESCRIPTION
**Pull request by LingaoM**
Since most of field in the pool should not change at runtime, So make this struct as rodata, and move variables to runtime ctx. Put this structure to rom look like more safely.

https://github.com/zephyrproject-rtos/zephyr/issues/66265

### CODE REVIEW

**Comment by LingaoM** [comment]
**west build -b bbc_microbit_v2 zephyr/samples/bluetooth/peripheral -p auto**

**before**
``` C
Memory region         Used Size  Region Size  %age Used
           FLASH:      194988 B       512 KB     37.19%
             RAM:       27133 B       128 KB     20.70%
        IDT_LIST:          0 GB         2 KB      0.00%
```

**after**
``` C
Memory region         Used Size  Region Size  %age Used
           FLASH:      195024 B       512 KB     37.20%
             RAM:       27021 B       128 KB     20.62%
        IDT_LIST:          0 GB         2 KB      0.00%
```

**Review by jukkar — COMMENTED** [review]
No objections to the change from my side, some nits about naming.

File: include/zephyr/net/buf.h [hunk 1]
```diff
@@ -1007,26 +1012,36 @@ struct net_buf_pool {
...
-#define NET_BUF_POOL_INITIALIZER(_pool, _alloc, _bufs, _count, _ud_size, _destroy) \
+#define _NET_BUF_POOL_RUNTIME_DEFINE(_name, _count)				\
+	static struct net_buf_pool_rt_ctx _net_buf_pool_rt##_name = {		\
+		.free = Z_LIFO_INITIALIZER((_net_buf_pool_rt##_name).free),	\
+		.lock = { },							\
+		.uninit_count = _count,						\
+		.avail_count = _count,						\
+	};
+
+#define NET_BUF_POOL_INITIALIZER(_pool, _alloc, _rt, _bufs, _count, _ud_size, _destroy) \
```
Comment (Thalley) [inline include/zephyr/net/buf.h:1023]: Similarly you are also here modifying a stable API by adding `_rt`.

**Review by Thalley — CHANGES_REQUESTED** [review]

File: include/zephyr/net/buf.h [hunk 2]
```diff
@@ -969,25 +969,30 @@ struct net_buf_data_alloc {
...
+		struct k_spinlock lock;
+
+		/** Number of uninitialized buffers */
+		uint16_t uninit_count;
+
+#if defined(CONFIG_NET_BUF_POOL_USAGE)
+		/** Amount of available buffers in the pool. */
+		atomic_t avail_count;
+#endif /* CONFIG_NET_BUF_POOL_USAGE */
+	} * const rt;
```
[the remark this answers is not in the document]
Comment (LingaoM) [reply include/zephyr/net/buf.h:987]: Are these structures really part of what is available to users?

[the remark this answers is not in the document]
Comment (Thalley) [reply include/zephyr/net/buf.h:987]: We can't know what applications are doing, but they certainly could. 

You even updated a babblesim test yourself to use this new API: https://github.com/zephyrproject-rtos/zephyr/pull/66250/files#diff-52d8c19d68f0679aa9dd25e434951a32c1d907315f5b96d1c4c855e420aa4fea

File: include/zephyr/net/buf.h (hunk 1 above)
Comment (LingaoM) [reply include/zephyr/net/buf.h:1026]: 😭

File: include/zephyr/net/buf.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (Thalley) [reply include/zephyr/net/buf.h:987]: You may want to get an review from @rerickson1 before changing anything :)

**Pull request closed without merging**